### PR TITLE
pkg/util/workqueue/prometheus: fix double registration

### LINKED ISSUE #76956: duplicate metrics collector registration attempted
**What happened**:

The following error messages were found in the Kubernetes API server and controller manager logs:

```
E0423 17:35:58.491576       1 prometheus.go:138] failed to register depth metric admission_quota_controller: duplicate metrics collector registration attempted
E0423 17:35:58.491625       1 prometheus.go:150] failed to register adds metric admission_quota_controller: duplicate metrics collector registration attempted
E0423 17:35:58.491901       1 prometheus.go:162] failed to register latency metric admission_quota_controller: duplicate metrics collector registration attempted
E0423 17:35:58.492191       1 prometheus.go:174] failed to register work_duration metric admission_quota_controller: duplicate metrics collector registration attempted
E0423 17:35:58.492367       1 prometheus.go:189] failed to register unfinished_work_seconds metric admission_quota_controller: duplicate metrics collector registration attempted
E0423 17:35:58.492507       1 prometheus.go:202] failed to register longest_running_processor_microseconds metric admission_quota_controller: duplicate metrics collector registration attempted
```

These appear to be for deprecated metrics (https://github.com/kubernetes/kubernetes/pull/71300) noted in https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.14.md.

**What you expected to happen**:

No ` duplicate metrics collector registration attempted` error messages in the API server or controller manager logs.

**How to reproduce it (as minimally and precisely as possible)**:

Deploy a Kubernetes version 1.14 cluster and review API server or controller manager logs.

**Anything else we need to know?**:

No.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.14.1
- Cloud provider or hardware configuration: IBM Cloud
- OS (e.g: `cat /etc/os-release`): Ubuntu 16.04.6 LTS
- Kernel (e.g. `uname -a`): 4.4.0-145-generic
- Install tools: Custom
- Network plugin and version (if this is a network-related bug): N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by s-urbaniak**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Currently, if workqueue metrics are registered twice, these metrics will be ignored.
This fixes it.

**Which issue(s) this PR fixes**:

Fixes #76956

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
/sig instrumentation
/priority backlog

**Comment by brancz** [comment]
Interesting solution, I do wonder, shouldn't these metrics be vector metrics? Then we wouldn't have this problem in the first place.

**Comment by s-urbaniak** [comment]
@brancz this is indeed a great idea :+1: let me play with that, now that this has a unit test.

**Review by logicalhan — COMMENTED** [review]
Have you verified that double registration renders the existing metrics defunct? I do not believe that that is actually the case (of course I could be wrong).

If you look in the [documentation for `prometheus.Desc`](https://github.com/prometheus/client_golang/blob/master/prometheus/desc.go#L28-L43), then you can see that descriptors are considered equal if the fully-qualified name and the labels are the same. In the case of the workqueue metrics, even if the function is invoked multiple times, each metric which is returned is therefore considered identical to the registry.

**Comment by s-urbaniak** [comment]
@logicalhan yes I verified this by looking the registration logic in the registry, it first checks if the collector is already registered in 

https://github.com/kubernetes/kubernetes/blob/e9af72c6e9524ed85c827298f87220e324444d36/vendor/github.com/prometheus/client_golang/prometheus/registry.go#L327-L332

and only sets it in the registry later if the above detection did not fail:
https://github.com/kubernetes/kubernetes/blob/e9af72c6e9524ed85c827298f87220e324444d36/vendor/github.com/prometheus/client_golang/prometheus/registry.go#L339-L347

and used a small verification snippet https://play.golang.org/p/WsnIqXbddzP

**Comment by s-urbaniak** [comment]
@brancz added vector metrics, where possible. The deprecated ones are not label based so I thing we have to use the `AlreadyRegisteredError` approach on these.

Once we have consensus on this PR, I'll address the bazel failures as they seem mechanical.

File: pkg/util/workqueue/prometheus/prometheus.go [hunk 1]
```diff
@@ -38,120 +38,156 @@ const (
...
+	// this should fail hard as this indicates a programmatic error, i.e.
+	// an invalid or duplicate metric descriptor,
+	// a previously registered descriptor with the same fqdn but different labels,
+	// or inconsistent label names or help strings for the same fqdn.
+	klog.Fatalf("failed to register metric %v name %v: %v", metric, name, err)
+	return nil
+}
+
 func (prometheusMetricsProvider) NewDeprecatedDepthMetric(name string) workqueue.GaugeMetric {
 	depth := prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (s-urbaniak) [inline pkg/util/workqueue/prometheus/prometheus.go]: nevermind, as discussed with @logicalhan we can go with simple singletons here and can omit mustRegister at all :+1: I will adapt the PR accordingly.

Comment (logicalhan) [reply pkg/util/workqueue/prometheus/prometheus.go]: Thanks!

**Review by logicalhan — COMMENTED** [review]
Thanks for taking the time on this!

File: pkg/util/workqueue/prometheus/prometheus.go (hunk 1 above)
Comment (s-urbaniak) [reply pkg/util/workqueue/prometheus/prometheus.go]: @logicalhan @brancz unfortunately declaring those deprecated metrics as singletons doesn't seem to work, as the subsystem is dynamic (sorry when I discussed this i was blind).

To omit the ugly mustRegister method, I added another commit which completely removes the deprecated metrics for good. @brancz the question is, if we can do this at this point.

**Comment by brancz** [comment]
Thanks for cleaning this up!

/lgtm

**Review by logicalhan — APPROVED** [review]
/lgtm

Looks good to me as well, thanks so much for fixing this!

**Comment by s-urbaniak** [comment]
ping @logicalhan @brancz Not sure what ` Not mergeable. Must be in milestone v1.15.` means. I guess this is just missing some label?

**Comment by brancz** [comment]
I think that was an outdated status. This now just needs an approver.

cc @smarterclayton

**Comment by s-urbaniak** [comment]
ping @smarterclayton for review and approval :-)

**Pull request merged**